### DISCUSSION TITLE
How to programmatically navigate to a new page without refreshing

### DISCUSSION THREAD

**Discussion by Prottoy2938** [vercel/next.js#11854] [Help]
I want to change my `URL` programmatically without losing my `state`. 

And by **changing URL**, I mean go from `/home` to `/about`.
I tried using `Shallow Routing`, but I don't seem to know how to properly implement it.

So how do I programmatically navigate to another `URL` without refreshing my page?

**Comment by ScreamZ** [comment]
You need to use whether [next links](https://nextjs.org/docs/api-reference/next/link) or [router.push](https://nextjs.org/docs/routing/imperatively).

If you need a global state maintained between pages (such as redux) use a [custom App component](https://nextjs.org/docs/advanced-features/custom-app) that wraps this.

**Reply by Prottoy2938** [reply]
I'm using this:
```
if (e.keyCode === 13) {
      e.preventDefault();
      const href = `/${encodeURIComponent(value)}`;  //value is from state
      Router.push(href, href, { shallow: true });
    }
```
But the page still refreshes during navigation

**Reply by Timer** [reply]
It is not possible to shallowly navigate between two different pages. Shallow navigation is for navigation within the same page.

If you need to save state across pages, you'll need to store it in a Custom App!

**Comment by alipetarian** [comment]
Perhaps you are looking for https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs/
